File: SudokuSolver_Java/src/main/sudokusolver/java/logic/simple/PointingPairsPointingTriples.java
package sudokusolver.java.logic.simple;

import sudokusolver.java.Board;
import sudokusolver.java.Cell;
import sudokusolver.java.LocatedCandidate;
import sudokusolver.java.RemoveCandidates;
import sudokusolver.java.SudokuNumber;
import sudokusolver.java.UnsolvedCell;

import java.util.Arrays;
import java.util.List;
import java.util.function.IntFunction;
import java.util.function.ToIntFunction;
import java.util.stream.Stream;

/*
 * https://www.sudokuwiki.org/Intersection_Removal#IR
 *
 * For a given block, if a candidate appears in only one row, then the candidate for that row must be placed in that
 * block. The candidate can be removed from cells which are in the same row, but different blocks.
 *
 * For a given block, if a candidate appears in only one column, then the candidate for that column must be placed in
 * that block. The candidate can be removed from cells which are in the same column, but different blocks.
 */
public class PointingPairsPointingTriples {
    public static List<RemoveCandidates> pointingPairsPointingTriples(Board<Cell> board) {
        return board.getBlocks()
                .stream()
                .flatMap(block -> Arrays.stream(SudokuNumber.values()).flatMap(candidate -> {
                    var rowModifications = pointingPairsPointingTriples(
                            block,
                            candidate,
                            board::getRow,
                            Cell::row
                    );
                    var columnModifications = pointingPairsPointingTriples(
                            block,
                            candidate,
                            board::getColumn,
                            Cell::column
                    );
                    return Stream.concat(rowModifications, columnModifications);
                }))
                .collect(LocatedCandidate.mergeToRemoveCandidates());
    }

    private static Stream<LocatedCandidate> pointingPairsPointingTriples(
            List<Cell> block,
            SudokuNumber candidate,
            IntFunction<List<Cell>> getUnit,
            ToIntFunction<Cell> getUnitIndex
    ) {
        var unitIndices = block.stream()
                .filter(UnsolvedCell.class::isInstance)
                .map(UnsolvedCell.class::cast)
                .filter(cell -> cell.candidates().contains(candidate))
                .mapToInt(getUnitIndex)
                .distinct()
                .toArray();
        if (unitIndices.length == 1) {
            return getUnit.apply(unitIndices[0])
                    .stream()
                    .filter(UnsolvedCell.class::isInstance)
                    .map(UnsolvedCell.class::cast)
                    .filter(cell -> cell.block() != block.getFirst().block() && cell.candidates().contains(candidate))
                    .map(cell -> new LocatedCandidate(cell, candidate));
        } else {
            return Stream.empty();
        }
    }
}

File: SudokuSolver_Kotlin/src/main/sudokusolver/kotlin/logic/PointingPairsPointingTriples.kt
package sudokusolver.kotlin.logic

import sudokusolver.kotlin.Board
import sudokusolver.kotlin.Cell
import sudokusolver.kotlin.RemoveCandidates
import sudokusolver.kotlin.SudokuNumber
import sudokusolver.kotlin.UnsolvedCell
import sudokusolver.kotlin.mergeToRemoveCandidates

/*
 * http://www.sudokuwiki.org/Intersection_Removal#IR
 *
 * For a given block, if a candidate appears in only one row, then the candidate for that row must be placed in that
 * block. The candidate can be removed from cells which are in the same row, but different blocks.
 *
 * For a given block, if a candidate appears in only one column, then the candidate for that column must be placed in
 * that block. The candidate can be removed from cells which are in the same column, but different blocks.
 *
 * For each block
 *   For each candidate
 *     If the candidate appears in unsolved cells of the same row
 *       Remove the candidate from cells of the row which are outside of the block
 *     If the candidate appears in unsolved cells of the same column
 *       Remove the candidate from cells of the column which are outside of the block
 */
fun pointingPairsPointingTriples(board: Board<Cell>): List<RemoveCandidates> =
    board.blocks.flatMap { block ->
        val unsolved = block.filterIsInstance<UnsolvedCell>()
        SudokuNumber.values().mapNotNull { candidate ->

            fun removeFromOtherBlocks(unit: List<Cell>) =
                unit.filterIsInstance<UnsolvedCell>()
                    .filter { it.block != block.first().block && candidate in it.candidates }
                    .map { it to candidate }

            val withCandidate = unsolved.filter { candidate in it.candidates }
            when {
                withCandidate.map { it.row }.toSet().size == 1 ->
                    removeFromOtherBlocks(board.getRow(withCandidate.first().row))

                withCandidate.map { it.column }.toSet().size == 1 ->
                    removeFromOtherBlocks(board.getColumn(withCandidate.first().column))

                else -> null
            }
        }.flatten()
    }.mergeToRemoveCandidates()